### PR TITLE
nextjs sample apps code

### LINKED ISSUE #50441: [SDK] make the sample apps for nextjs that are using the compatibility layer

### PR DESCRIPTION
**Pull request by npretto**
Closes https://github.com/metabase/metabase/issues/50441

# Description

This is the code for the nextjs sample apps.
The main branch has the "clean" apps that were created from `create-next-app`, so this PR only includes the changes from the generated apps.

The PR repo incudes both a pages router and a app router app.
Both require [next sdk compatibility layer](https://github.com/metabase/metabase/pull/50230), this is why in the package.json the dependency is using a link, we don't have a version number to use yet. I'll update that when we publish the first version that has the next-js compatibility layer.

# How to test

Clone this repo in the same folder of your metabase repo, then cd in the repo folder and for each app:
- cp .env.sample .env
- npm install
- npm run dev

if you need a quick script to nuke node_modules and next cache, copy the envs and install deps for both projects:
```
cd next-sample-app-router
rm -rf node_modules
rm -rf .next
cp .env.sample .env
npm install --install-links ../../metabase/resources/embedding-sdk

cd ../next-sample-pages-router
rm -rf node_modules
rm -rf .next
cp .env.sample .env
npm install --install-links ../../metabase/resources/embedding-sdk
```

The two apps will open on (by default) ports 3001 and 3002

Demo:

https://github.com/user-attachments/assets/425fdf61-4bd2-4bf3-bac4-05e87c13f65f

### CODE REVIEW

**Comment by oisincoveney** [comment]
@npretto would it make sense to add the script that you put in the description as a script in the PR? something like `start-all` or something like that

**Comment by npretto** [comment]
> @npretto would it make sense to add the script that you put in the description as a script in the PR? something like `start-all` or something like that

That script is mostly to clean caches, which we only need when changing sdk versions, real users will most likely only cd the folder they want (either the app or pages router) and do the usual `npm install; npm run dev`, wdyt?

**Comment by oisincoveney** [comment]
> > @npretto would it make sense to add the script that you put in the description as a script in the PR? something like `start-all` or something like that
> 
> That script is mostly to clean caches, which we only need when changing sdk versions, real users will most likely only cd the folder they want (either the app or pages router) and do the usual `npm install; npm run dev`, wdyt?

yeah fair enough

**Review by oisincoveney — APPROVED** [review]

**Comment by albertoperdomo** [comment]
Tested locally. Looks good to me.
Let's get the compatibility layer merged and release to npm first, then update the PR to install the SDK from npm instead of filesystem.

**Comment by WiNloSt** [comment]
For some reason, I couldn't test the page (:3001) router, and I was greeted with this error message.
image

The app (:3002) router works fine though 🤔 
image

I used https://github.com/metabase/metabase/pull/50230 to build the sdk with `yarn build-embedding-sdk`. And I tried `rm -rf node_modules && npm install && npm run dev` for both routers just to be safe.

**Comment by npretto** [comment]
> For some reason, I couldn't test the page (:3001) router, and I was greeted with this error message. image
> 
> The app (:3002) router works fine though 🤔 image
> 
> I used [metabase/metabase#50230](https://github.com/metabase/metabase/pull/50230) to build the sdk with `yarn build-embedding-sdk`. And I tried `rm -rf node_modules && npm install && npm run dev` for both routers just to be safe.

Ouch, I don't know why that's coming up only on the app router but I know how to fix it 🫠 
I'll add it to the docs later today, the fix should be to use `npm install --install-links`
Otherwise npm will use symlinks and sometimes that makes the app find two reacts, one in the main app, and a different one for the sdk.
This should affect only developers building locally, so I'll add it to https://www.metabase.com/docs/latest/developers-guide/sdk

File: next-sample-pages-router/src/pages/api/metabase/auth.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    return res.end();
+  } catch (error) {
+    console.log("error", error);
+    if (error instanceof Error) {
+      res.write(error.message);
+      return res.end();
+    }
+    res.write("unknown error");
+    return res.end();
+  }
```
Comment (heypoom) [inline next-sample-pages-router/src/pages/api/metabase/auth.ts:57]: Could we use `res.send(session)` as it is equivalent to write and end?

**Review by heypoom — APPROVED** [review]
Nice!

File: next-sample-pages-router/src/pages/api/metabase/auth.ts (hunk 1 above)
Comment (npretto) [reply next-sample-pages-router/src/pages/api/metabase/auth.ts:57]: I'll add it in a follow up pr so I won't need to re-request for review on this one, we'll have to do one anyway after we merge the renaming `config -> authConfig`

Comment (heypoom) [reply next-sample-pages-router/src/pages/api/metabase/auth.ts:57]: @npretto sounds good! I think there's `defineEmbeddingSdkConfig -> defineMetabaseAuthConfig` too

**Pull request merged**